tests/unit/test_offerendpoint.py

```diff
@@ -0,0 +1,63 @@
+#
+# Test our offer endpoint parser
+#
+
+import unittest
+
+from juju.offerendpoints import (OfferEndpoints, OfferURL, ParseError,
+                                 maybe_parse_offer_url_source,
+                                 parse_offer_endpoint, parse_offer_url)
+
+
+class TestOfferEndpoint(unittest.TestCase):
+
+    def test_parse_offer_endpoint(self):
+
+        cases = {"mysql:db": OfferEndpoints("mysql", ["db"]),
+                 "model/fred.mysql:db": OfferEndpoints("mysql", ["db"], "model", "model/fred"),
+                 "mysql:db,log": OfferEndpoints("mysql", ["db", "log"]),
+                 }
+        for name, case in cases.items():
+            self.assertEqual(parse_offer_endpoint(name), case)
+
+
+class TestOfferURL(unittest.TestCase):
+
+    def test_parse_offer_url(self):
+        cases = {"controller:user/modelname.applicationname": OfferURL(source="controller", user="user", model="modelname", application="applicationname"),
+                 "controller:user/modelname.applicationname:rel": OfferURL(source="controller", user="user", model="modelname", application="applicationname:rel"),
+                 "modelname.applicationname": OfferURL(model="modelname", application="applicationname"),
+                 "modelname.applicationname:rel": OfferURL(model="modelname", application="applicationname:rel"),
+                 "/modelname.applicationname": OfferURL(model="modelname", application="applicationname"),
+                 "/modelname.applicationname:rel": OfferURL(model="modelname", application="applicationname:rel"),
+                 "user/modelname.applicationname": OfferURL(user="user", model="modelname", application="applicationname"),
+                 "user/modelname.applicationname:rel": OfferURL(user="user", model="modelname", application="applicationname:rel"),
+                 }
+        for name, case in cases.items():
+            self.assertEqual(parse_offer_url(name), case)
+
+    def test_parse_offer_url_failures(self):
+        cases = {"controller:modelname": "application offer URL is missing application",
+                 "controller:user/modelname": "application offer URL is missing application",
+                 "modelname": "application offer URL is missing application",
+                 "/user/modelname": "application offer URL is missing application",
+                 "modelname.applicationname@bad": "application name applicationname@bad not valid",
+                 "user[bad/model.application": "user name user[bad not valid",
```

Comment:
are these brackets right? user[bad ? with no close?

Reply:
Yeah, it just an invalid character in the username